src/common/stack_mmap.c

```diff
@@ -0,0 +1,422 @@
+		/* new stack size has been added to the btree */
+		sp->nb_sizes++;
+		d_list_add_tail(&(*spb)->size_list, &sp->stack_size_list);
+		D_DEBUG(DB_MEM, "sub-pool by-size %p has been created in pool %p for size %zu\n",
+			*spb, sp, size);
+	} else {
+		/* this should not happen, size finally exists! so no need new/same one */
+		D_DEBUG(DB_MEM,
+			"sub-pool by-size %p of size %zu already exists in pool %p finally..., so freeing %p\n",
+			*((void **)item), size, sp, *spb);
+		D_FREE(*spb);
+		*spb = *((void **)item);
+	}
+
+	return 0;
+}
+
+int mmap_stack_thread_create_common(struct stack_pool *sp, enum AbtThreadCreateType flag,
+				    void *arg, void (*thread_func)(void *), void *thread_arg,
+				    ABT_thread_attr attr, ABT_thread *newthread)
+{
+	ABT_thread_attr new_attr = ABT_THREAD_ATTR_NULL;
+	int rc;
+	void *stack;
+	mmap_stack_desc_t *mmap_stack_desc = NULL;
+	size_t stack_size = MMAPED_ULT_STACK_SIZE, usable_stack_size;
+	struct stack_pool_by_size *spb = NULL;
+
+	if (daos_ult_mmap_stack == false) {
+		rc = call_abt_method(arg, flag, thread_func, thread_arg,
+				     attr, newthread);
+		return rc;
+	}
+
+	if (attr != ABT_THREAD_ATTR_NULL) {
+		ABT_thread_attr_get_stack(attr, &stack, &stack_size);
+		if (stack != NULL) {
+			/* an other external stack allocation method is being
+			 * used, nothing to do
+			 */
+			rc = call_abt_method(arg, flag, thread_func, thread_arg,
+					     attr, newthread);
+			return rc;
+		}
+		if (stack_size < MMAPED_ULT_STACK_SIZE)
+			stack_size = MMAPED_ULT_STACK_SIZE;
+	} else {
+		rc = ABT_thread_attr_create(&new_attr);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("Create ABT thread attr failed: %d\n", rc);
+			return rc;
+		}
+		attr = new_attr;
+	}
+
+	/* XXX a stack is allocated from the creating XStream's stack pool
+	 * but will be freed on the running one ...
+	 */
+
+	rc = stack_pool_by_size_find_or_create(sp, &spb, stack_size);
+	if (rc != 0) {
+		D_ERROR("unable to find/create stack sub-pool in pool %p for size %zu : %d\n",
+			sp, stack_size, rc);
+		/* let's try Argobots standard way ... */
+		rc = call_abt_method(arg, flag, thread_func, thread_arg,
+				     attr, newthread);
+		D_GOTO(out_err, rc);
+	}
+
+	if ((mmap_stack_desc = d_list_pop_entry(&spb->stack_free_list,
+						mmap_stack_desc_t,
+						stack_list)) != NULL) {
+		D_ASSERT(sp->free_stacks != 0);
+		--sp->free_stacks;
+		atomic_fetch_sub(&nb_free_stacks, 1);
+		stack = mmap_stack_desc->stack;
+		stack_size = mmap_stack_desc->stack_size;
+		D_DEBUG(DB_MEM,
+			"mmap()'ed stack %p of size %zd from free list, in pool=%p / sub-pool=%p, remaining free stacks in pool="DF_U64", on CPU=%d\n",
+			stack, stack_size, sp, spb, sp->free_stacks, sched_getcpu());
+	} else {
+		/* XXX this test is racy, but if max_nb_mmap_stacks value is
+		 * high enough it does not matter as we do not expect so many
+		 * concurrent ULTs creations during mmap() syscall to cause
+		 * nb_mmap_stacks to significantly exceed max_nb_mmap_stacks ...
+		 */
```

Comment:
I think all the constraints should be per 'stack_pool'(which is per-xstream for DAOS), instead of per engine. You could move them into the 'stack_pool' data structure, and 'atomic' operations could be avoided.

Reply:
I mean shouldn't we split limit and assign them to all xstreams equally? (like what we did for other server resources) Do we really need all xstreams share the limit?